sdk/core/Azure.Core.Experimental/src/MutableJsonElement.WriteTo.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Text.Json;
+
+namespace Azure.Core.Json
+{
+    public partial struct MutableJsonElement
+    {
+        internal void WriteTo(Utf8JsonWriter writer)
+        {
+            WriteElement(_path, _highWaterMark, _element, writer);
+            writer.Flush();
+        }
+
+        private void WriteElement(string path, int highWaterMark, JsonElement originalElement, Utf8JsonWriter writer)
+        {
+            if (Changes.TryGetChange(path, highWaterMark, out MutableJsonChange change))
+            {
+                change.AsJsonElement().WriteTo(writer);
+                return;
+            }
+
+            if (Changes.DescendantChanged(path, highWaterMark))
+            {
+                switch (originalElement.ValueKind)
+                {
+                    case JsonValueKind.Object:
+                        WriteObject(path, highWaterMark, originalElement, writer);
+                        break;
+                    case JsonValueKind.Array:
+                        WriteArray(path, highWaterMark, originalElement, writer);
+                        break;
+                    default:
+                        throw new InvalidOperationException("Element doesn't have descendants.");
+                }
+
+                return;
+            }
+
+            originalElement.WriteTo(writer);
+        }
+
+        private void WriteObject(string path, int highWaterMark, JsonElement originalElement, Utf8JsonWriter writer)
+        {
+            writer.WriteStartObject();
+
+            foreach (JsonProperty property in originalElement.EnumerateObject())
```

Comment:
this enumerates original properties. What if a new property got added?

Reply:
Good catch!  Fixing...

Reply:
Oh wait, no, that was a copy and paste error in my test.  I think it does work, but let me come up with a clear explanation, and I can update the variable names to clarify as well.

Reply:
Ok, I've added tests for this [here](https://github.com/Azure/azure-sdk-for-net/pull/35204/files#diff-64c6d61f33cd5e11ec907e4f27f4b455c2434e70079d4f3f720d03c2e68dbe92R255) and [here](https://github.com/Azure/azure-sdk-for-net/pull/35204/files#diff-64c6d61f33cd5e11ec907e4f27f4b455c2434e70079d4f3f720d03c2e68dbe92R279).

The explanation for this is that when a property is added to an object, we [store a change to that object's path](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.Experimental/src/MutableJsonElement.cs#L310-L317) with the new JsonElement holding the added property.  So, we enter [the first block in WriteElement](https://github.com/Azure/azure-sdk-for-net/pull/35204/files#diff-0b846adb4112f2d61c0ffb34cb025964feba66e862cd55e6e052e43f578dd256R19), and write out the entire property in the block.

I'll add some more tests with different sequencings of changes to make the test coverage a little more robust.